docs/rfcs/0002-session-id-sharing.md

```diff
@@ -0,0 +1,226 @@
+
+#### Main application
+
+```cpp
+MainApp::init() {
+  String lingeringSessionId = LocalStorage::get(SESSION_ID_STORAGE_KEY);
+  // If this is not empty, the VPN was not deactivated from the app.
+  if (!lingeringSessionId.isEmpty()) {
+    if (Daemon::vpnIsOn()) {
+      try {
+        String sessionId = Daemon::getSessionId();
+        // Check if we are still in the same session
+        if (sessionId != lingeringSessionId) {
+          // Flush any lingering data from a previous session
+          pings::main_app::session.submit(INIT_FLUSH_REASON);
+
+          metrics::main_app::sessionId.set(sessionId);
+          // Record session::* context metrics e.g. app_exclusions.
+
+          LocalStorage::set(SESSION_ID_STORAGE_KEY, sessionId);
+        }
+
+        // Schedule the timer.
+        MainApp::schedulePingTimer();
+      } catch {
+        // Note: If communication with the daemon doesn't work,
+        // let's assume the VPN is off. This is very unlikely, given we
+        // have just communicated with the daemon through Daemon::vpnIsOn();
+
+        // Flush any lingering data from a previous session
+        pings::main_app::session.submit(INIT_FLUSH_REASON);
+        // Clear the sessionId storage
+        LocalStorage::clear(SESSION_ID_STORAGE_KEY);
+      }
+    } else {
+      // Flush any lingering data from a previous session
+      pings::main_app::session.submit(INIT_FLUSH_REASON);
+      // Clear the sessionId storage
+      LocalStorage::clear(SESSION_ID_STORAGE_KEY);
+    }
+  }
+}
+
+MainApp::onActivate() {
+  // Let's flush any lingering data for before activating the VPN.
+  // We want the active session pings to only include data from the current session.
+  //
+  // This ping should be always empty. This is a way to monitor if it isn't
+  pings::main_app::session.submit(FLUSH_REASON);
+
+  String sessionId = generateUUID();
+  try {
+    Daemon::activate(config, sessionId);
+    // Note: session::* metrics should only be recorded from now on,
+    // so that flush pings don't contain unexpected data.
+
+    metrics::main_app::sessionId.set(sessionId);
+    // Record session::* context metrics e.g. app_exclusions.
+
+    LocalStorage::set(SESSION_ID_STORAGE_KEY, sessionId);
+
+    // Submit the start ping
+    pings::main_app::session.submit(START_REASON);
+  } catch {
+    // TODO: Record error.
+    LocalStorage::clear(SESSION_ID_STORAGE_KEY);
+  }
+}
+
+MainApp::onPingTimer() {
+  String sessionId = LocalStorage::get(SESSION_ID_STORAGE_KEY);
+  metrics::main_app::sessionId.set(sessionId);
+
+  // Submit the timer ping
+  pings::main_app::session.submit(TIMER_REASON);
+}
+
+MainApp::onDeactivate() {
+  // Stop the timer.
+  MainApp::stopPingTimer();
+
+  String sessionId = LocalStorage::get(SESSION_ID_STORAGE_KEY);
+  metrics::main_app::sessionId.set(sessionId);
+
+  // Submit the stop ping
+  pings::main_app::session.submit(TIMER_REASON);
+  LocalStorage::clear(SESSION_ID_STORAGE_KEY);
+}
+```
+
+#### Daemon
```

Comment:
Can you share some psuedocode for iOS? I don't think this will apply over there.

Reply:
I'd rather keep the pseudo-code generic. Do you have specific concerns for iOS? I'd rather go over them explicitly.

Reply:
On iOS it's not a daemon, and so there isn't a concept of `init`. Where would that `init` block of psuedo code be on iOS? I think that information should be included here. (The other three daemon blocks have obvious analogous functions on iOS.) You're right, we don't need iOS-specific psuedocode overall, but it should be clear where the psuedocode will be implemented on both platforms - and I'm not sure it is for this bit.

Reply:
Yes, there is a concept of init. See: https://github.com/mozilla-mobile/mozilla-vpn-client/blob/main/src/apps/vpn/platforms/ios/iostunnel.swift#L17-L41. The terminology is different on iOS, but the network extension is just a separate process running some code.